### ISSUE TITLE
@OnNeverAskAgain not being called

### ISSUE THREAD

**Issue by emarc-m** [hotchemi/PermissionsDispatcher#78]
Annotating a method with @OnNeverAskAgain is not called. What is called is the method annotated with @OnPermissionDenied.

I think the generated code conditions checks are incorrect.

```java
static void onRequestPermissionsResult(BaseActivity target, int requestCode, int[] grantResults) {
    switch (requestCode) {
      case REQUEST_INITIALIZESCANNING:
      // this would always be true if the permission was denied and on never ask again
      if (!PermissionUtils.hasSelfPermissions(target, PERMISSION_INITIALIZESCANNING)) {
        target.showDenied();
        return;
      }
      if (PermissionUtils.verifyPermissions(grantResults)) {
        target.initializeScanning();
      } else {
        if (!PermissionUtils.shouldShowRequestPermissionRationale(target, PERMISSION_INITIALIZESCANNING)) {
          target.showNeverAsk(); // this is NEVER called
        } else {
          target.showDenied();
        }
      }
      break;
      default:
      break;
    }
  }
```

Thanks. Great work BTW.

**Comment by hotchemi** [comment]
@emarc-m Ah, could u use 2.0.2 for the time being? It might works well. So sorry!

**Comment by emarc-m** [comment]
@hotchemi Thanks for that info.

**Comment by shiraji** [comment]
It's time to add functional tests.

**Comment by hotchemi** [comment]
Right code is bellow.

```diff
static void onRequestPermissionsResult(BaseActivity target, int requestCode, int[] grantResults) {
    switch (requestCode) {
      case REQUEST_INITIALIZESCANNING:
      // this would always be true if the permission was denied and on never ask again
+     if (Build.VERSION. SDK_INT < 23 && !PermissionUtils.hasSelfPermissions(target, PERMISSION_INITIALIZESCANNING)) {
        target.showDenied();
        return;
      }
      if (PermissionUtils.verifyPermissions(grantResults)) {
        target.initializeScanning();
      } else {
        if (!PermissionUtils.shouldShowRequestPermissionRationale(target, PERMISSION_INITIALIZESCANNING)) {
          target.showNeverAsk(); // this is NEVER called
        } else {
          target.showDenied();
        }
      }
      break;
      default:
      break;
    }
  }
```

**Comment by hotchemi** [comment]
@shiraji Might be the good idea but it takes a bunch of costs...worrisome problem.

**Comment by shiraji** [comment]
Wait, I think there is one more problem. PermissionDispatcher must take care of SYSTEM_ALERT_WINDOW and WRITE_SETTINGS. 

This [test case](https://github.com/hotchemi/PermissionsDispatcher/blob/master/processor/src/test/java/permissions/dispatcher/processor/data/Source.java#L5171) is not correct. Should be following:

```java
if (!Settings.System.canWrite(target)) 
```

This [line](https://github.com/hotchemi/PermissionsDispatcher/pull/72/files#diff-f3a29e581c90035d2a26d60d08e2bc94R275) should check if it is in `ADD_WITH_CHECK_BODY_MAP`

We need to refactor `fun addHasSelfPermissionsCondition(builder: MethodSpec.Builder, activityVar: String)` as well in order to take care of this one.

Do you want me to send pull request for this?

**Comment by hotchemi** [comment]
@shiraji Umm, actually this line is just workaround for under M devices. See https://github.com/hotchemi/PermissionsDispatcher/issues/45#issuecomment-160058769 .

I'm not sure the same thing happens in SYSTEM_ALERT_WINDOW and WRITE_SETTINGS.
So, we should delete this line when exceptional permissions case.

> Do you want me to send pull request for this?
Sure!

**Comment by shiraji** [comment]
> So, we should delete this line when exceptional permissions case.

That makes life easier!

**Comment by hotchemi** [comment]
@emarc-m Just released 2.0.4. Really appreciated for your report.

**Comment by emarc-m** [comment]
@hotchemi Thanks for the support.
